sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3963,8 +3966,10 @@ object TimeWindowing extends Rule[LogicalPlan] {
 
         def getWindow(i: Int, dataType: DataType): Expression = {
           val timestamp = PreciseTimestampConversion(window.timeColumn, dataType, LongType)
-          val lastStart = timestamp - (timestamp - window.startTime
-            + window.slideDuration) % window.slideDuration
+          val remainder = (timestamp - window.startTime) % window.slideDuration
+          val lastStart = CaseWhen(
+            Seq((LessThan(remainder, 0), timestamp - remainder - window.slideDuration))
```

Comment:
Thank you for reviewing.

For any event timestamp, the start time of the last window (whether tumbling or sliding) obtained by it should always be less than the current timestamp. When `(timestamp - window.starttime)% window.slideduration <0`, the obtained laststart will be greater than timestamp. At this time, it is necessary to shift the window to the right, which is the correct laststart value.

For example

code
```
    val timestamp = -13
    val offset = 0
    val windowSize = 7

    // old code
    val lastStartOld = timestamp - (timestamp - offset - windowSize) % windowSize

   // new code
    val remainder =  (timestamp - offset) % windowSize

    val lastStartNew =
      if (remainder < 0) {
        timestamp - remainder - windowSize
      } else {
        timestamp - remainder - windowSize
      }

    println(s"lastStartOld = $lastStartOld   lastStartNew = $lastStartNew")
```
result
```
lastStartOld = -7 lastStartNew = -14
```

Reply:
I think the question is how this arises in the first place. The code is self explanatory, not asking you to explain it

Reply:
@srowen sure,As mentioned above, I think this problem is unavoidable because we can't control the behavior of users. All we can do is get the right results even if users enter illegal data.